server/src-lib/Hasura/GraphQL/Schema/Remote.hs

```diff
@@ -32,41 +36,56 @@ buildRemoteParser
   -> m ( [P.FieldParser n RemoteField]
        , Maybe [P.FieldParser n RemoteField]
        , Maybe [P.FieldParser n RemoteField])
-buildRemoteParser (IntrospectionResult sdoc query_root mutation_root subscription_root) info = do
-  queryT <- makeParsers query_root
-  mutationT <- traverse makeParsers mutation_root
-  subscriptionT <- traverse makeParsers subscription_root
+buildRemoteParser (IntrospectionResult sdoc queryRoot mutationRoot subscriptionRoot) info = do
+  queryT <- makeParsers queryRoot
+  mutationT <- makeNonQueryRootFieldParser mutationRoot $$(G.litName "Mutation")
+  subscriptionT <- makeNonQueryRootFieldParser subscriptionRoot $$(G.litName "Subscription")
   return (queryT, mutationT, subscriptionT)
   where
-    makeFieldParser :: G.FieldDefinition -> m (P.FieldParser n RemoteField)
+    makeFieldParser :: RemoteSchemaFieldDefinition -> m (P.FieldParser n RemoteField)
     makeFieldParser fieldDef = do
       fldParser <- remoteField' sdoc fieldDef
-      pure $ (info, ) <$> fldParser
+      pure $ (RemoteField info) <$> fldParser
     makeParsers :: G.Name -> m [P.FieldParser n RemoteField]
     makeParsers rootName =
       case lookupType sdoc rootName of
         Just (G.TypeDefinitionObject o) ->
           traverse makeFieldParser $ _otdFieldsDefinition o
         _ -> throw400 Unexpected $ rootName <<> " has to be an object type"
 
+    -- | The spec says that the `schema` definition can be omitted, if the root names
+    --   are the defaults (Query, Mutation and Subscription). This function is used
+    --   to constructor a `FieldParser` for the mutation and subscription roots.
+    --   If the user has given a custom Mutation/Subscription root name, then it will
+    --   look for that and if it's not found in the schema document, then an error is thrown.
+    --   If no root name has been provided, we lookup the schema document for an object with
+    --   the default name and if that's not found, we omit the said Root from the schema.
+    makeNonQueryRootFieldParser :: Maybe G.Name -> G.Name -> m (Maybe [P.FieldParser n RemoteField])
```

Comment:
This definitely feels like an isolated feature / bug fix that should be own PR?

Reply:
Well, this is not a bug until this feature because when we do an introspection query to a GraphQL server, it includes the root names. The default root names can only come through the GraphQL IDL.

Yes, Indeed it is an isolated feature. I will be happy to raise the PR for the same. But, the problem with this being in a separate PR is that there'll be no way to test this (I think).